Sources/AsyncDNSResolver/dnssd/DNSResolver_dnssd.swift

```diff
@@ -138,32 +125,42 @@ struct DNSSD {
                 handler.handleRecord(errorCode: errorCode, data: rdata, length: rdlen)
             }
 
-            let serviceRefPtr = UnsafeMutablePointer<DNSServiceRef?>.allocate(capacity: 1)
-            defer { serviceRefPtr.deallocate() }
-
             // Run the query
             let _code = DNSServiceQueryRecord(
-                serviceRefPtr,
+                query.serviceRefPtr,
                 kDNSServiceFlagsTimeout,
                 0,
                 name,
                 UInt16(type.kDNSServiceType),
                 UInt16(kDNSServiceClass_IN),
                 callback,
-                handlerPointer
+                query.replyHandlerPointer
             )
 
             // Check if query completed successfully
             guard _code == kDNSServiceErr_NoError else {
                 return continuation.finish(throwing: AsyncDNSResolver.Error(dnssdCode: _code))
             }
 
-            // Read reply from the socket (blocking) then call reply handler
-            DNSServiceProcessResult(serviceRefPtr.pointee)
-            DNSServiceRefDeallocate(serviceRefPtr.pointee)
+            let serviceSockFD = DNSServiceRefSockFD(query.serviceRefPtr.pointee)
+            guard serviceSockFD != -1 else {
+                return continuation.finish(throwing: AsyncDNSResolver.Error(code: .internalError, message: "Failed to access the DNSSD service socket"))
+            }
+
+            let readSource = DispatchSource.makeReadSource(fileDescriptor: serviceSockFD)
+            readSource.setEventHandler {
+                // Read reply from the socket (blocking) then call reply handler
+                DNSServiceProcessResult(query.serviceRefPtr.pointee)
 
-            // Streaming done
-            continuation.finish()
+                // Streaming done
+                continuation.finish()
+            }
+            readSource.resume()
+
+            continuation.onTermination = { _ in
+                readSource.cancel()
+                DNSServiceRefDeallocate(query.serviceRefPtr.pointee)
```

Comment:
security relevant use after free bug here, you could be concurrently deallocating this as well as using it in the event handler

Reply:
The deallocation is now done in the dispatch source cancel handler so that should no longer happen